src/host/kvpair.rs

```diff
@@ -388,33 +388,24 @@ impl<const DEPTH: usize> MerkleTree<[u8; 32], DEPTH> for MongoMerkle<DEPTH> {
     }
 
     fn get_node_with_hash(&self, index: u32, hash: &[u8; 32]) -> Result<Self::Node, MerkleError> {
-        let v = self.get_record(index, hash).expect("Unexpected DB Error");
         let height = (index + 1).ilog2();
-        v.map_or_else(
-            || {
-                let default = self.get_default_hash(height as usize)?;
-                let child_hash = if height == Self::height() as u32 {
-                    [0; 32]
-                } else {
-                    self.get_default_hash((height + 1) as usize)?
-                };
-                if default == *hash {
-                    Ok(MerkleRecord {
-                        index,
-                        hash: self.get_default_hash(height as usize)?,
-                        data: [0; 32],
-                        left: child_hash,
-                        right: child_hash,
-                    })
-                } else {
-                    Err(MerkleError::new(*hash, index, MerkleErrorCode::InvalidHash))
-                }
-            },
-            |x| {
-                assert!(x.index == index);
-                Ok(x)
-            },
-        )
+        let default = self.get_default_hash(height as usize)?;
+        let child_hash = if height == Self::height() as u32 {
+            [0; 32]
+        } else {
+            self.get_default_hash((height + 1) as usize)?
+        };
+        if default == *hash {
+            Ok(MerkleRecord {
+                index,
+                hash: default,
+                data: [0; 32],
+                left: child_hash,
+                right: child_hash,
+            })
+        } else {
+            Err(MerkleError::new(*hash, index, MerkleErrorCode::InvalidHash))
```

Comment:
Not very understanding this changing.
It looks like change the logic to if the input hash is not default one, it will return error.
This looks like not correct? As after leaf changed, the parent nodes can change to non-default hash. 
This is the reason why we need get it from cache/db first. If we cannot find it in db, it means the node was not touch so it must be default.

Do I miss understanding anything?

Reply:
Seems the name of get_node_with_hash is not a correct name for the same logic.